### PR TITLE
2.3 release.py sftp retries

### PR DESCRIPTION
**Pull request by mumrah**
Add retry capability to the `cmd` function in release.py. This allows for selectively retrying certain commands which might be flaky, like the SFTP puts.

### CODE REVIEW

**Comment by mumrah** [comment]
Trying this for RC1 of 2.3.1, saw this during most recent run:

```
Uploading artifacts in /Users/david.arthur/Code/Confluent/prod/kafka/.release_work_dir/kafka-2.3.1-rc1/javadoc/org/apache/kafka/streams/kstream to your Apache home directory [u'sftp', u'-b', u'-', u'davidarthur@home.apache.org'] --> 
put /Users/david.arthur/Code/Confluent/prod/kafka/.release_work_dir/kafka-2.3.1-rc1/javadoc/org/apache/kafka/streams/kstream/package-frame.html public_html/kafka-2.3.1-rc1/javadoc/org/apache/kafka/streams/kstream/package-frame.html

> ssh: connect to host home.apache.org port 22: No route to host
> Connection closed
> 
Retrying... 2 remaining retries
Uploading artifacts in /Users/david.arthur/Code/Confluent/prod/kafka/.release_work_dir/kafka-2.3.1-rc1/javadoc/org/apache/kafka/streams/kstream to your Apache home directory [u'sftp', u'-b', u'-', u'davidarthur@home.apache.org'] 
> sftp> 
> sftp> put /Users/david.arthur/Code/Confluent/prod/kafka/.release_work_dir/kafka-2.3.1-rc1/javadoc/org/apache/kafka/streams/kstream/package-frame.html public_html/kafka-2.3.1-rc1/javadoc/org/apache/kafka/streams/kstream/package-frame.html
> 
```

and the script continues on.

File: release.py [hunk 1]
```diff
@@ -96,10 +96,12 @@ def print_output(output):
...
 
-def cmd(action, cmd, *args, **kwargs):
-    if isinstance(cmd, basestring) and not kwargs.get("shell", False):
-        cmd = cmd.split()
+def cmd(action, cmd_arg, *args, **kwargs):
+    if isinstance(cmd_arg, basestring) and not kwargs.get("shell", False):
+        cmd_arg = cmd_arg.split()
+
     allow_failure = kwargs.pop("allow_failure", False)
+    num_retries = kwargs.pop("num_retries", 0)
```
Comment (ewencp) [inline release.py:104]: What kinds of failures are we trying to mask here? Frequently retries also require delays between retries.

File: release.py [hunk 2]
```diff
@@ -109,16 +111,23 @@ def cmd(action, cmd, *args, **kwargs):
...
+        output = subprocess.check_output(cmd_arg, *args, stderr=subprocess.STDOUT, **kwargs)
         print_output(output)
     except subprocess.CalledProcessError as e:
         print_output(e.output)
 
+        if num_retries > 0:
+            kwargs['num_retries'] = num_retries - 1
+            kwargs['allow_failure'] = allow_failure
+            print("Retrying... %d remaining retries" % (num_retries - 1))
+            return cmd(action, cmd_arg, *args, **kwargs)
```
Comment (ewencp) [inline release.py:125]: Probably fine currently given max number of retries we specify, but would really be preferable to do this iteratively (which also doesn't require copying/mgmt of kwargs like this).

**Review by ewencp — APPROVED** [review]
LGTM, just a few minor questions.

**Comment by ijuma** [comment]
@mumrah you may want to get this into trunk before the upcoming release.

**Comment by mumrah** [comment]
Closing in favor of #8021 which is based on trunk

File: release.py (hunk 1 above)
Comment (mumrah) [reply release.py:104]: Added a sleep in the new PR

**Pull request closed without merging**